x/evm/keeper/context_stack.go

```diff
@@ -0,0 +1,76 @@
+package keeper
+
+import (
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+type cachedContext struct {
+	ctx    sdk.Context
+	commit func()
+}
+
+// ContextStack manages the initial context and a stack of cached contexts,
+// to support the `StateDB.Snapshot` and `StateDB.RevertToSnapshot` methods.
+type ContextStack struct {
+	// Context of the initial state before transaction execution.
+	// It's the context used by `StateDB.CommitedState`.
+	initialCtx     sdk.Context
+	cachedContexts []cachedContext
+}
+
+// CurrentContext returns the top context of cached stack,
+// if the stack is empty, returns the initial context.
+func (k *ContextStack) CurrentContext() sdk.Context {
+	l := len(k.cachedContexts)
+	if l == 0 {
+		return k.initialCtx
+	}
+	return k.cachedContexts[l-1].ctx
+}
+
+// Reset sets the initial context and clear the cache context stack.
+func (k *ContextStack) Reset(ctx sdk.Context) {
```

Comment:
hmmm this look more like a `NewContextStack` than a reset

Maybe it would be more clear to rename `RevertAll` to `Reset` instead and this one to `NewContextStack` 

@fedekunze any opinion on this?

Reply:
yes, this looks like a constructor function. Let's rename it to `NewContextStack`

Reply:
The only downside of that style is that the `EvmKeeper.WithContext` will need to construct a new object everytime.